### PR TITLE
feat: convert `schedule.channel` (char field) to `schedule.slack_channel` (foreign key)

### PR DESCRIPTION
**Pull request by joeyorlando**
# What this PR does

`OnCallSchedule` equivalent of https://github.com/grafana/oncall/pull/5191.

**NOTE**: merge after https://github.com/grafana/oncall/pull/5224 (so that I can use some of the new serializer fields defined in there)

### Migration
```bash
Running migrations:                                                                                                                                                                                                │
│ source=engine:app google_trace_id=none logger=apps.schedules.migrations.0019_auto_20241021_1735 Starting migration to populate slack_channel field.                                                                │
│ source=engine:app google_trace_id=none logger=apps.schedules.migrations.0019_auto_20241021_1735 Total schedules to process: 1                                                                                      │
│ source=engine:app google_trace_id=none logger=apps.schedules.migrations.0019_auto_20241021_1735 Schedule 26 updated with SlackChannel 2 (slack_id: C043LL6RTS7).                                                   │
│ source=engine:app google_trace_id=none logger=apps.schedules.migrations.0019_auto_20241021_1735 Bulk updated 1 OnCallSchedules with their Slack channel.                                                           │
│ source=engine:app google_trace_id=none logger=apps.schedules.migrations.0019_auto_20241021_1735 Finished migration. Total schedules processed: 1. Schedules updated: 1. Missing SlackChannels: 0.                  │
│   Applying schedules.0019_auto_20241021_1735... OK
```

### Tested Public API
```txt
POST {{oncall_host}}/api/v1/schedules/
Authorization: {{oncall_api_key}}
Content-Type: application/json

{
    "name": "Demo testy testy2",
    "type": "web",
    "time_zone": "America/Los_Angeles",
    "slack": {
        "channel_id": "C05PPLYN1U1"
    }
}

HTTP/1.1 201 Created
Content-Type: application/json
Vary: Accept, Origin
Allow: GET, POST, HEAD, OPTIONS
X-Frame-Options: DENY
Content-Length: 198
X-Content-Type-Options: nosniff
Referrer-Policy: same-origin
Cross-Origin-Opener-Policy: same-origin

{
  "id": "SBBN73UTUTVCE",
  "team_id": null,
  "name": "Demo testy testy2",
  "time_zone": "America/Los_Angeles",
  "on_call_now": [],
  "shifts": [],
  "slack": {
    "channel_id": "C05PPLYN1U1",
    "user_group_id": null
  },
  "type": "web"
}
```

### Tested via UI (eg; internal API)

https://www.loom.com/share/e66bf3468b144dd782da5eb6e0bfd0af

## Checklist

- [x] Unit, integration, and e2e (if applicable) tests updated
- [x] Documentation added (or `pr:no public docs` PR label added if not required)
- [x] Added the relevant release notes label (see labels prefixed w/ `release:`). These labels dictate how your PR will
    show up in the autogenerated release notes.

### CODE REVIEW

File: engine/apps/api/serializers/organization.py [hunk 1]
```diff
@@ -27,7 +27,7 @@ class OrganizationSerializer(EagerLoadingMixin, serializers.ModelSerializer):
     rbac_enabled = serializers.BooleanField(read_only=True, source="is_rbac_permissions_enabled")
     grafana_incident_enabled = serializers.BooleanField(read_only=True, source="is_grafana_incident_enabled")
 
-    SELECT_RELATED = ["slack_team_identity"]
+    SELECT_RELATED = ["slack_team_identity", "slack_channel"]
```
Comment (joeyorlando) [inline engine/apps/api/serializers/organization.py:30]: forgot to include this in https://github.com/grafana/oncall/pull/5191

File: engine/apps/public_api/serializers/schedules_base.py
Comment (joeyorlando) [inline engine/apps/public_api/serializers/schedules_base.py:1]: TODO: merge this PR after https://github.com/grafana/oncall/pull/5224, this way I can refactor this to use from `common.api_helpers.custom_fields.SlackChannelsFilteredByOrganizationSlackWorkspaceField` (to further deduplicate the validation stuff in this serializer)

File: engine/apps/schedules/models/on_call_schedule.py [hunk 2]
```diff
@@ -207,7 +208,16 @@ class OnCallSchedule(PolymorphicModel):
...
     name = models.CharField(max_length=200)
+
+    # TODO: drop this field in a subsequent release, this has been migrated to slack_channel field
     channel = models.CharField(max_length=100, null=True, default=None)
+    slack_channel = models.ForeignKey(
+        "slack.SlackChannel",
+        null=True,
+        default=None,
+        on_delete=models.SET_NULL,
+        related_name="+",
```
Comment (vadimkerr) [inline engine/apps/schedules/models/on_call_schedule.py:219]: what does `related_name="+"` do? is it some kind of alias for smth?

**Review by vadimkerr — APPROVED** [review]
LGTM

File: engine/apps/schedules/models/on_call_schedule.py (hunk 2 above)
Comment (joeyorlando) [reply engine/apps/schedules/models/on_call_schedule.py:219]: Screenshot 2024-11-04 at 12 13 30 PM

so basically for a `SlackChannel` instance, there won't be a `schedules` attribute to interact w/ all on-call schedules associated w/ that slack channel (probably not needed for now, but we can change in the future if need be).

File: engine/apps/public_api/serializers/schedules_base.py
Comment (joeyorlando) [reply engine/apps/public_api/serializers/schedules_base.py:1]: done 👍

File: engine/apps/public_api/serializers/schedules_base.py [hunk 3]
```diff
@@ -1,83 +1,66 @@
...
+    def to_representation(self, instance: "OnCallSchedule"):
+        return {
+            **super().to_representation(instance),
+            "slack": {
+                "channel_id": instance.slack_channel_slack_id,
+                "user_group_id": instance.user_group.slack_id if instance.user_group is not None else None,
+            },
         }
 
-        return result
```
Comment (joeyorlando) [inline engine/apps/public_api/serializers/schedules_base.py:79]: see `SlackSerializer` above.. all of the validation logic is deduplicated (and tested) in those custom serializers

**Pull request merged**